### PR TITLE
Allow unicode and nobigmath hints to be used with site data

### PR DESCRIPTION
**Pull request by FherStk**
A new field called "binary data" has been added to the "problem_data" model, so this field will be stored into the "init.yaml" file when storing the test cases for a problem.

### CODE REVIEW

**Review by FherStk — COMMENTED** [review]
Hints also added (code including accents will work properly).

File: judge/migrations/0134_problemdata_hints.py [hunk 1]
```diff
@@ -0,0 +1,18 @@
+# Generated by Django 3.2.16 on 2022-12-12 16:25
+
+from django.db import migrations, models
+
+
+class Migration(migrations.Migration):
+
+    dependencies = [
+        ('judge', '0133_auto_20221212_1427'),
```
Comment (kiritofeng) [inline judge/migrations/0134_problemdata_hints.py:9]: You'll need to rename this now

File: judge/models/problem_data.py [hunk 2]
```diff
@@ -30,6 +30,11 @@ def problem_directory_file(data, filename):
     ('linecount', _('Line-by-line')),
 )
 
+HINTS = (
+    ('unicode', _('Unicode')),
+    ('none', _('NONE'))
```
Comment (kiritofeng) [inline judge/models/problem_data.py:35]: We also have `nobigmath` (https://github.com/DMOJ/java-sandbox-agent); `nobuf` isn't necessary because there's a language-agnostic `unbuffered` option that should be used instead.

File: judge/migrations/0133_add_binary_data.py [hunk 3]
```diff
@@ -0,0 +1,457 @@
...
+import django.core.validators
+import django.db.models.deletion
+from django.db import migrations, models
+
+import judge.models.contest
+import judge.models.problem
+import judge.models.profile
+
+
+class Migration(migrations.Migration):
```
Comment (kiritofeng) [inline judge/migrations/0133_add_binary_data.py:12]: @Xyene @quantum5 we need to do the cleanup migration at some point, should we write a separate migration or just hide it here?

File: judge/migrations/0134_problemdata_hints.py (hunk 1 above)
Comment (FherStk) [reply judge/migrations/0134_problemdata_hints.py]: Done!

**Comment by Xyene** [comment]
Thanks for the contribution. While I'm generally in favour of exposing this functionality in the site interface, I'd rather we didn't expose it as `hints` directly. I think the implementation of hints on the judge side was a mistake and these should have just been fields, e.g.,

```yaml
require_unicode_output: true
disable_bigmath_libraries: true
```

instead of the more arcane

```yaml
hints: [unicode, nobigmath]
```

We might still go back and clean this up on the judge side at some point.

With that in mind, I think we should be exposing three boolean fields in this PR, and only internally outputting the hints field.

**Comment by quantum5** [comment]
Yes, these should be boolean fields. Also, since binary data does not solve your problem, I would rather that one is not exposed unless the need truly arises.

**Comment by FherStk** [comment]
Pull request updated :)

File: .gitignore [hunk 4]
```diff
@@ -10,3 +10,8 @@ dmoj/local_settings.py
 resources/style.css
 resources/martor-description.css
 sass_processed
+dmojsite
+dmojsite/*
+problems
+problems/*
+*.log
```
Comment (quantum5) [inline .gitignore:19]: These lines seem specific to your install. It's recommended to not store unrelated files inside the repository anyways. Would remove this from the PR.

**Comment by FherStk** [comment]
Sorry, I forgot to pull from master, my branch is now updated and so the pull request.

File: judge/models/problem_data.py [hunk 5]
```diff
@@ -91,4 +92,4 @@ class ProblemTestCase(models.Model):
     output_limit = models.IntegerField(verbose_name=_('output limit length'), blank=True, null=True)
     checker = models.CharField(max_length=10, verbose_name=_('checker'), choices=CHECKERS, blank=True)
     checker_args = models.TextField(verbose_name=_('checker arguments'), blank=True,
-                                    help_text=_('Checker arguments as a JSON object.'))
+                                    help_text=_('Checker arguments as a JSON object.'))
```
Comment (quantum5) [inline judge/models/problem_data.py:95]: Can you add the trailing `\n` so that git doesn't complain?

File: judge/utils/problem_data.py [hunk 6]
```diff
@@ -178,10 +180,18 @@ def make_checker(case):
...
+        if self.data.unicode:
+            hints.append("unicode")            
+        if self.data.nobigmath:
+            hints.append("nobigmath")                     
         if self.data.checker:
             init['checker'] = make_checker(self.data)
         else:
             self.data.checker_args = ''
+
+        if len(hints) > 0:
```
Comment (quantum5) [inline judge/utils/problem_data.py:192]:
```suggestion
        if hints:
```

**Review by quantum5 — COMMENTED** [review]
Can we squash all the migrations together? Otherwise LGTM except for the minor nits.

File: .gitignore (hunk 4 above)
Comment (FherStk) [reply .gitignore:19]: Sure! Removing.

File: judge/models/problem_data.py (hunk 5 above)
Comment (FherStk) [reply judge/models/problem_data.py:95]: Sorry, do you mean at the end of the file?

Comment (quantum5) [reply judge/models/problem_data.py:95]: Yes.

File: judge/utils/problem_data.py (hunk 6 above)
Comment (FherStk) [reply judge/utils/problem_data.py:192]: Done!

File: judge/models/problem_data.py (hunk 5 above)
Comment (FherStk) [reply judge/models/problem_data.py:95]: Done!

**Comment by FherStk** [comment]
PR updated :)

File: .gitignore [hunk 7]
```diff
@@ -11,4 +11,4 @@ resources/dark
 resources/martor-description.css
 resources/style.css
 resources/vars.scss
-sass_processed
+sass_processed
```
Comment (quantum5) [inline .gitignore:14]: Can we not change this file at all?

File: judge/migrations/0133_add_binary_data.py [hunk 8]
```diff
@@ -0,0 +1,457 @@
+# Generated by Django 3.2.16 on 2022-12-12 14:27
```
Comment (quantum5) [inline judge/migrations/0133_add_binary_data.py:1]: Can we squash all three migrations?

**Review by quantum5 — COMMENTED** [review]
Thanks for making the changes! Unfortunately, there are some linter errors and the migrations still need to be squashed.

File: .gitignore (hunk 7 above)
Comment (FherStk) [reply .gitignore:14]: I apologize, downloading the original file again.

File: judge/migrations/0133_add_binary_data.py (hunk 8 above)
Comment (FherStk) [reply judge/migrations/0133_add_binary_data.py:1]: Sorry, it was my first squash. Done!

**Comment by FherStk** [comment]
PR updated again, thanks @quantum5 for your infinite patience with me :)

**Review by quantum5 — COMMENTED** [review]
Changes look good, but you are still failing the linter. Otherwise it looks good.

**Comment by FherStk** [comment]
I'm not sure about how to proceed, should I fix this?
image

**Comment by quantum5** [comment]
> I'm not sure about how to proceed, should I fix this? image

Yes. Once you are done, you should run flake8 locally to make sure all errors disappear.

**Comment by FherStk** [comment]
Done, I've installed fake8 locally with all the needed plugins: no errors on my local side.

**Review by quantum5 — APPROVED** [review]

File: judge/migrations/0133_add_binary_data.py (hunk 3 above)
Comment (FherStk) [reply judge/migrations/0133_add_binary_data.py]: Done!
